src/content/articles/for-in-string-types.mdx

```diff
@@ -0,0 +1,212 @@
+As a result, when you write a `for...in` loop over the keys of an object, you often won't have a guarantee that the object's keys will only be from the type of the object.
+Some other area of code might have provided an object with more properties that just so happens to match the structural type.
+
+Take a look at this completely type safe code snippet that passes an object to a function.
+What types would you expect for `i` inside the function's `for` loop?
+
+```ts twoslash
+interface TwoCounts {
+  one: number;
+  two: number;
+}
+
+function iterateOverCounts(counts: TwoCounts) {
+  for (const i in counts) {
+    // ...
+  }
+}
+
+const counts = {
+  one: 1,
+  two: 2,
+  three: 3,
+};
+
+iterateOverCounts(counts);
+```
+
+`i` at _runtime_ will be `"one"`, `"two"`, then `"three"`.
+But there's no reasonable way for TypeScript's type system to know that.
+It just knows that `counts` is type `TwoCounts`, which has keys `"one"` and `"two"`.
+
+Thus, if TypeScript tried to assume that `i` was `keyof typeof counts` then the type system could be insidiously wrong.
+For example, your code might assume `i` is only ever `"one"` or `"two`", and use it in logic that would crash with any other value!
+
+See [Workarounds](#workarounds) below for how you can get around this behavior in `for...in` loops.
+
+:::tip
+Wondering why the first code snippet couldn't narrow `counts` to a more precise type?
+See also [Why Objects Aren't Narrowed Before Usage](/articles/objects-narrowed-before-usage).
+:::
+
+### `Object.keys()` Too
+
+This same design quirk applies to `Object.keys()`:
+
+```ts twoslash
+const counts = {
+  one: 1,
+  two: 2,
+};
+
+Object.keys(counts);
+//     ^?
+```
+
+`Object.keys()` and `for...in` loops work almost the same way.
+The main difference is that `Object.keys()` only looks at properties on the object itself, while `for...in` loops also look at properties on the object's `prototype` chain.
+
+See [Workarounds](#workarounds) below for how you can get around this behavior in `Object.keys()` as well.
+
+## Workarounds
+
+Before you try to work around the design of the type checker, please pause a moment and reflect.
+The situations mentioned earlier in this post are real problems in JavaScript and TypeScript code that have caused countless bugs in software.
+
+**TypeScript is right to assume `string`s types.**
+
+### `Object` APIs
+
+TypeScript's type checking is designed to work with how JavaScript data typically flows and changes through an application.
+Needing to wrestle with the type checker is often a byproduct of application code going against safe JavaScript practices -- such as the aforementioned issues with `for...in` loops receiving unexpected key strings.
+In many cases, it might be good to listen to TypeScript and consider refactoring your code to use a JavaScript API better suited to the task.
+
+If all your code needs to do is get all the values in an object, an `Object.values()` API exists that returns those as an array.
```

Comment:
[Proofreading/nitpick/question] I edited the sentence a bit to make it clearer. Is the below suggestion the message you want to convey? 
```suggestion
If all your code needs all an object's values, an `Object.values()` API exists to return those as an array.
```